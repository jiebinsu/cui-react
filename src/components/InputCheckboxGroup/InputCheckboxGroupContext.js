import { createContext, useContext } from "react";

export const CheckboxContext = createContext({
  inGroup: false,
});

export const useCheckboxContext = () => useContext(CheckboxContext);
