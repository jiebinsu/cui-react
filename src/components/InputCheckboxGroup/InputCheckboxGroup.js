import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import uuid from "uuid";
import { CheckboxContext } from "./InputCheckboxGroupContext";
import "./inputCheckboxGroup.scss";
import classNames from "../../utils/classNames";

const InputCheckboxGroup = ({
  id,
  legend,
  hint,
  hasError,
  errorMsg,
  className,
  value,
  children,
  onChange,
  ...props
}) => {
  const [selfVal, setSelfVal] = useState(value);

  const updateState = (val, checked) => {
    const removed = selfVal.filter((v) => v !== val);
    const next = checked ? [...removed, val] : removed;
    setSelfVal(next);
    if (onChange) onChange(next);
  };

  useEffect(() => {
    setSelfVal(value);
  }, [value.join(",")]);

  const providerValue = {
    updateState,
    values: selfVal,
    inGroup: true,
  };

  const hintId = `${id}-hint`;
  const errorId = `${id}-error`;
  const classes = classNames("coop-c-form-choice", [
    hasError && "coop-form__invalid",
    className,
  ]);

  const ariaDescribedBy = [];
  if (hint) ariaDescribedBy.push(hintId);
  if (hasError) ariaDescribedBy.push(errorId);

  return (
    <CheckboxContext.Provider value={providerValue}>
      <div className="coop-form__row">
        <fieldset
          className={classes}
          aria-describedby={ariaDescribedBy && ariaDescribedBy.join(" ")}
          {...props}
        >
          <legend className="coop-form__legend coop-c-form-choice__legend">
            {legend}
          </legend>
          {hint && (
            <p id={`${id}-hint`} className="coop-form__hint">
              {hint}
            </p>
          )}
          {hasError && (
            <p id={`${id}-error`} className="coop-form__error">
              {errorMsg}
            </p>
          )}
          {children}
        </fieldset>
      </div>
    </CheckboxContext.Provider>
  );
};

export default InputCheckboxGroup;

InputCheckboxGroup.defaultProps = {
  id: uuid(),
  hasError: false,
  hint: null,
  errorMsg: null,
  children: null,
  className: null,
  value: [],
  onChange: null,
};

InputCheckboxGroup.propTypes = {
  id: PropTypes.string,
  legend: PropTypes.string.isRequired,
  hint: PropTypes.string,
  hasError: PropTypes.bool,
  errorMsg: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  value: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  ),
  onChange: PropTypes.func,
};
