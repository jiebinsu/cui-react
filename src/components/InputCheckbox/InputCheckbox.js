import React, { useState, useEffect, useCallback, forwardRef } from "react";
import PropTypes from "prop-types";
import { useCheckboxContext } from "../InputCheckboxGroup/InputCheckboxGroupContext";
import classNames from "../../utils/classNames";
import "./inputCheckbox.scss";

const InputCheckbox = forwardRef(
  (
    {
      id,
      name,
      className,
      label,
      checked,
      value: checkboxValue,
      disabled,
      onChange,
      ...props
    },
    ref
  ) => {
    const [selfChecked, setSelfChecked] = useState(!!checked);
    const { updateState, inGroup, values } = useCheckboxContext();
    const classes = classNames(
      "coop-form__field coop-form__checkbox coop-c-form-choice__input coop-c-form-choice__input--checkbox",
      [className]
    );

    if (inGroup) {
      useEffect(() => {
        const next = values.includes(checkboxValue);
        if (next === selfChecked) return;
        setSelfChecked(next);
      }, [values.join(",")]);
    }

    const onChangeHandler = useCallback(
      (event) => {
        if (disabled) return;

        const selfEvent = {
          target: {
            checked: !selfChecked,
          },
          stopPropagation: event.stopPropagation,
          preventDefault: event.preventDefault,
          nativeEvent: event,
        };
        if (inGroup && updateState) {
          if (updateState) updateState(checkboxValue, !selfChecked);
        }

        setSelfChecked(!selfChecked);
        if (onChange) onChange(selfEvent);
      },
      [updateState, onChange, disabled, selfChecked]
    );

    useEffect(() => {
      if (checked === undefined) return;
      setSelfChecked(!!checked);
    }, [checked]);

    const tagAttributes = {
      id,
      name,
      value: checkboxValue,
      label,
      disabled,
      type: "checkbox",
      className: classes,
      checked: selfChecked,
      onChange: onChangeHandler,
      ...props,
      ref,
    };

    return (
      <div className="coop-c-form-choice">
        <input {...tagAttributes} />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor={id}
        >
          {label}
        </label>
      </div>
    );
  }
);

InputCheckbox.defaultProps = {
  disabled: false,
  className: null,
  checked: null,
  onChange: null,
  value: "",
};

InputCheckbox.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  className: PropTypes.string,
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
};

export default InputCheckbox;
