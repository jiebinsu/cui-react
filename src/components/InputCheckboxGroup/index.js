import InputCheckboxGroup from "./InputCheckboxGroup";

export default InputCheckboxGroup;
