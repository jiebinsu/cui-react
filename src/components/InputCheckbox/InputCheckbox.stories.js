import React from "react";
import InputCheckbox from "./InputCheckbox";

export default {
  title: "Base/InputCheckbox",
  component: InputCheckbox,
};

export const Basic = (args) => <InputCheckbox {...args} />;
Basic.args = {
  id: "basic",
  name: "basic",
  label: "Basic Checkbox",
  value: 1,
};
