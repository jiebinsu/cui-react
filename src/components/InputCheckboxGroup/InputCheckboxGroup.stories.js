import React from "react";
import InputCheckboxGroup from "./InputCheckboxGroup";
import InputCheckbox from "../InputCheckbox";

export default {
  title: "Base/InputCheckboxGroup",
  component: InputCheckboxGroup,
};

export const CheckboxExample = () => (
  <>
    <InputCheckboxGroup
      id="group-1"
      legend="Checkbox Group with hint"
      hint="Hint message"
    >
      <InputCheckbox
        id="checkbox-1"
        name="checkbox-1"
        label="Checkbox 1"
        value={1}
      />
      <InputCheckbox
        id="checkbox-2"
        name="checkbox-2"
        label="Checkbox 2"
        value={2}
      />
      <InputCheckbox
        id="checkbox-3"
        name="checkbox-3"
        label="Checkbox 3"
        value={3}
      />
    </InputCheckboxGroup>
    <InputCheckboxGroup
      id="group-2"
      legend="Checkbox Group with hint and error"
      hint="Hint message"
      errorMsg="Error message"
      hasError
    >
      <InputCheckbox
        id="checkbox-4"
        name="checkbox-4"
        label="Checkbox 4"
        value={4}
      />
      <InputCheckbox
        id="checkbox-5"
        name="checkbox-5"
        label="Checkbox 5"
        value={5}
      />
      <InputCheckbox
        id="checkbox-6"
        name="checkbox-6"
        label="Checkbox 6"
        value={6}
      />
    </InputCheckboxGroup>
  </>
);
