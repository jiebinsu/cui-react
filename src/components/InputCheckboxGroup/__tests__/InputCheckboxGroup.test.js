import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import InputCheckboxGroup from "../index";
import InputCheckbox from "../../InputCheckbox";

describe("InputCheckboxGroup", () => {
  it("should render correctly", () => {
    const wrapper = mount(
      <InputCheckboxGroup
        id="group-1"
        legend="Checkbox Group with hint"
        hint="Hint message"
      />
    );
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should support hints and errors", () => {
    const All = () => (
      <div>
        <InputCheckboxGroup
          id="group-1"
          legend="Checkbox Group with hint"
          hint="Hint message"
        >
          <InputCheckbox
            id="checkbox-1"
            name="checkbox-1"
            label="Checkbox 1"
            value={1}
          />
          <InputCheckbox
            id="checkbox-2"
            name="checkbox-2"
            label="Checkbox 2"
            value={2}
          />
          <InputCheckbox
            id="checkbox-3"
            name="checkbox-3"
            label="Checkbox 3"
            value={3}
          />
        </InputCheckboxGroup>
        <InputCheckboxGroup
          id="group-2"
          legend="Checkbox Group with hint and error"
          hint="Hint message"
          errorMsg="Error message"
          hasError
        >
          <InputCheckbox
            id="checkbox-4"
            name="checkbox-4"
            label="Checkbox 4"
            value={4}
          />
          <InputCheckbox
            id="checkbox-5"
            name="checkbox-5"
            label="Checkbox 5"
            value={5}
          />
          <InputCheckbox
            id="checkbox-6"
            name="checkbox-6"
            label="Checkbox 6"
            value={6}
          />
        </InputCheckboxGroup>
      </div>
    );

    const tree = renderer.create(<All />).toJSON();
    expect(tree).toMatchSnapshot();

    const wrapper = mount(<All />);
    expect(() => wrapper.unmount()).not.toThrow();
  });
});
