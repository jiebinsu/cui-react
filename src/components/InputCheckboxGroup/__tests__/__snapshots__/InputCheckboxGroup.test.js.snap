// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputCheckboxGroup should support hints and errors 1`] = `
<div>
  <div
    className="coop-form__row"
  >
    <fieldset
      aria-describedby="group-1-hint"
      className="coop-c-form-choice"
    >
      <legend
        className="coop-form__legend coop-c-form-choice__legend"
      >
        Checkbox Group with hint
      </legend>
      <p
        className="coop-form__hint"
        id="group-1-hint"
      >
        Hint message
      </p>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__checkbox coop-c-form-choice__input coop-c-form-choice__input--checkbox"
          disabled={false}
          id="checkbox-1"
          label="Checkbox 1"
          name="checkbox-1"
          onChange={[Function]}
          type="checkbox"
          value={1}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="checkbox-1"
        >
          Checkbox 1
        </label>
      </div>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__checkbox coop-c-form-choice__input coop-c-form-choice__input--checkbox"
          disabled={false}
          id="checkbox-2"
          label="Checkbox 2"
          name="checkbox-2"
          onChange={[Function]}
          type="checkbox"
          value={2}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="checkbox-2"
        >
          Checkbox 2
        </label>
      </div>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__checkbox coop-c-form-choice__input coop-c-form-choice__input--checkbox"
          disabled={false}
          id="checkbox-3"
          label="Checkbox 3"
          name="checkbox-3"
          onChange={[Function]}
          type="checkbox"
          value={3}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="checkbox-3"
        >
          Checkbox 3
        </label>
      </div>
    </fieldset>
  </div>
  <div
    className="coop-form__row"
  >
    <fieldset
      aria-describedby="group-2-hint group-2-error"
      className="coop-c-form-choice coop-form__invalid"
    >
      <legend
        className="coop-form__legend coop-c-form-choice__legend"
      >
        Checkbox Group with hint and error
      </legend>
      <p
        className="coop-form__hint"
        id="group-2-hint"
      >
        Hint message
      </p>
      <p
        className="coop-form__error"
        id="group-2-error"
      >
        Error message
      </p>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__checkbox coop-c-form-choice__input coop-c-form-choice__input--checkbox"
          disabled={false}
          id="checkbox-4"
          label="Checkbox 4"
          name="checkbox-4"
          onChange={[Function]}
          type="checkbox"
          value={4}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="checkbox-4"
        >
          Checkbox 4
        </label>
      </div>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__checkbox coop-c-form-choice__input coop-c-form-choice__input--checkbox"
          disabled={false}
          id="checkbox-5"
          label="Checkbox 5"
          name="checkbox-5"
          onChange={[Function]}
          type="checkbox"
          value={5}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="checkbox-5"
        >
          Checkbox 5
        </label>
      </div>
      <div
        className="coop-c-form-choice"
      >
        <input
          checked={false}
          className="coop-form__field coop-form__checkbox coop-c-form-choice__input coop-c-form-choice__input--checkbox"
          disabled={false}
          id="checkbox-6"
          label="Checkbox 6"
          name="checkbox-6"
          onChange={[Function]}
          type="checkbox"
          value={6}
        />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor="checkbox-6"
        >
          Checkbox 6
        </label>
      </div>
    </fieldset>
  </div>
</div>
`;
